File: packages/policy-kit/src/evaluate.ts
/**
 * PEAC Policy Kit Evaluation
 *
 * Deterministic policy evaluation for CAL semantics.
 * First-match-wins rule semantics.
 *
 * @packageDocumentation
 */

import {
  PolicyDocument,
  PolicyRule,
  SubjectMatcher,
  EvaluationContext,
  EvaluationResult,
  ControlPurpose,
  ControlLicensingMode,
  SubjectType,
} from './types';

/**
 * Check if a value matches a single-or-array pattern
 *
 * @param value - Value to check
 * @param pattern - Single value or array of values
 * @returns true if value matches pattern
 */
function matchesSingleOrArray<T>(value: T | undefined, pattern: T | T[] | undefined): boolean {
  // If no pattern specified, match anything
  if (pattern === undefined) {
    return true;
  }

  // If no value and pattern exists, no match
  if (value === undefined) {
    return false;
  }

  // Check against array or single value
  if (Array.isArray(pattern)) {
    return pattern.includes(value);
  }

  return value === pattern;
}

/**
 * Check if a subject ID matches a pattern
 *
 * Supports:
 * - Exact match: "user:abc123"
 * - Prefix match with wildcard: "user:*" matches "user:abc123"
 *
 * @param id - Subject ID to check
 * @param pattern - Pattern to match against
 * @returns true if ID matches pattern
 */
function matchesIdPattern(id: string | undefined, pattern: string | undefined): boolean {
  // No pattern = match anything
  if (pattern === undefined) {
    return true;
  }

  // No ID but pattern exists = no match
  if (id === undefined) {
    return false;
  }

  // Wildcard prefix match
  if (pattern.endsWith('*')) {
    const prefix = pattern.slice(0, -1);
    return id.startsWith(prefix);
  }

  // Exact match
  return id === pattern;
}

/**
 * Check if subject labels contain all required labels
 *
 * @param subjectLabels - Labels on the subject
 * @param requiredLabels - Labels required by the rule
 * @returns true if subject has all required labels
 */
function hasAllLabels(
  subjectLabels: string[] | undefined,
  requiredLabels: string[] | undefined
): boolean {
  // No required labels = match
  if (requiredLabels === undefined || requiredLabels.length === 0) {
    return true;
  }

  // Required labels but no subject labels = no match
  if (subjectLabels === undefined || subjectLabels.length === 0) {
    return false;
  }

  // Check all required labels are present
  return requiredLabels.every((label) => subjectLabels.includes(label));
}

/**
 * Check if a subject matches a subject matcher
 *
 * @param subject - Subject from evaluation context
 * @param matcher - Subject matcher from rule
 * @returns true if subject matches all criteria
 */
function matchesSubject(
  subject: EvaluationContext['subject'],
  matcher: SubjectMatcher | undefined
): boolean {
  // No matcher = match any subject
  if (matcher === undefined) {
    return true;
  }

  // Check type
  if (!matchesSingleOrArray(subject?.type, matcher.type)) {
    return false;
  }

  // Check labels (must have ALL required labels)
  if (!hasAllLabels(subject?.labels, matcher.labels)) {
    return false;
  }

  // Check ID pattern
  if (!matchesIdPattern(subject?.id, matcher.id)) {
    return false;
  }

  return true;
}

/**
 * Check if a rule matches the evaluation context
 *
 * All criteria must match (AND logic).
 *
 * @param rule - Policy rule to check
 * @param context - Evaluation context
 * @returns true if rule matches context
 */
function ruleMatches(rule: PolicyRule, context: EvaluationContext): boolean {
  // Check subject
  if (!matchesSubject(context.subject, rule.subject)) {
    return false;
  }

  // Check purpose
  if (!matchesSingleOrArray(context.purpose, rule.purpose)) {
    return false;
  }

  // Check licensing mode
  if (!matchesSingleOrArray(context.licensing_mode, rule.licensing_mode)) {
    return false;
  }

  return true;
}

/**
 * Evaluate a policy against a context
 *
 * Uses first-match-wins semantics:
 * - Rules are evaluated in order
 * - First matching rule determines the decision
 * - If no rule matches, defaults are applied
 *
 * @param policy - Policy document
 * @param context - Evaluation context
 * @returns Evaluation result
 */
export function evaluate(policy: PolicyDocument, context: EvaluationContext): EvaluationResult {
  // Find first matching rule
  for (const rule of policy.rules) {
    if (ruleMatches(rule, context)) {
      return {
        decision: rule.decision,
        matched_rule: rule.name,
        reason: rule.reason,
        is_default: false,
      };
    }
  }

  // No rule matched, apply defaults
  return {
    decision: policy.defaults.decision,
    reason: policy.defaults.reason,
    is_default: true,
  };
}

/**
 * Explain which rules could potentially match a context
 *
 * Useful for debugging and policy analysis.
 * Returns all rules that would match if evaluated, in order.
 *
 * @param policy - Policy document
 * @param context - Evaluation context
 * @returns Array of rule names that match, or 'default' if none
 */
export function explainMatches(policy: PolicyDocument, context: EvaluationContext): string[] {
  const matches: string[] = [];

  for (const rule of policy.rules) {
    if (ruleMatches(rule, context)) {
      matches.push(rule.name);
    }
  }

  if (matches.length === 0) {
    matches.push('[default]');
  }

  return matches;
}

/**
 * Find the effective rule for a context
 *
 * Same as evaluate() but returns the full rule object.
 *
 * @param policy - Policy document
 * @param context - Evaluation context
 * @returns Matched rule or undefined if default applies
 */
export function findEffectiveRule(
  policy: PolicyDocument,
  context: EvaluationContext
): PolicyRule | undefined {
  for (const rule of policy.rules) {
    if (ruleMatches(rule, context)) {
      return rule;
    }
  }
  return undefined;
}

/**
 * Check if a policy would allow a given context
 *
 * Convenience helper for common allow/deny checks.
 *
 * @param policy - Policy document
 * @param context - Evaluation context
 * @returns true if decision is 'allow'
 */
export function isAllowed(policy: PolicyDocument, context: EvaluationContext): boolean {
  const result = evaluate(policy, context);
  return result.decision === 'allow';
}

/**
 * Check if a policy would deny a given context
 *
 * Convenience helper for common allow/deny checks.
 *
 * @param policy - Policy document
 * @param context - Evaluation context
 * @returns true if decision is 'deny'
 */
export function isDenied(policy: PolicyDocument, context: EvaluationContext): boolean {
  const result = evaluate(policy, context);
  return result.decision === 'deny';
}

/**
 * Check if a policy requires review for a given context
 *
 * Convenience helper for review checks.
 *
 * @param policy - Policy document
 * @param context - Evaluation context
 * @returns true if decision is 'review'
 */
export function requiresReview(policy: PolicyDocument, context: EvaluationContext): boolean {
  const result = evaluate(policy, context);
  return result.decision === 'review';
}

/**
 * Batch evaluate multiple contexts against a policy
 *
 * Useful for testing or bulk authorization checks.
 *
 * @param policy - Policy document
 * @param contexts - Array of evaluation contexts
 * @returns Array of evaluation results (same order as contexts)
 */
export function evaluateBatch(
  policy: PolicyDocument,
  contexts: EvaluationContext[]
): EvaluationResult[] {
  return contexts.map((context) => evaluate(policy, context));
}

File: sdks/go/policy/evaluate.go
package policy

import (
	"strings"
)

// ReasonNilPolicy is the reason returned when evaluating a nil policy.
const ReasonNilPolicy = "nil policy"

// Evaluate evaluates a policy against a context and returns the result.
// Rules are evaluated in order; the first matching rule wins.
// If no rule matches, the default decision is used.
//
// If policy is nil, returns a deny result with reason ReasonNilPolicy.
// If context is nil, an empty context is used.
func Evaluate(policy *PolicyDocument, context *EvaluationContext) *EvaluationResult {
	// Guard against nil policy
	if policy == nil {
		return &EvaluationResult{
			Decision:  Deny,
			Reason:    ReasonNilPolicy,
			IsDefault: true,
		}
	}

	if context == nil {
		context = &EvaluationContext{}
	}

	// Evaluate rules in order - first match wins
	for _, rule := range policy.Rules {
		if ruleMatches(&rule, context) {
			return &EvaluationResult{
				Decision:    rule.Decision,
				MatchedRule: rule.Name,
				Reason:      rule.Reason,
				IsDefault:   false,
			}
		}
	}

	// No rule matched, use defaults
	result := &EvaluationResult{
		Decision:  Deny, // Default to deny if no defaults specified
		IsDefault: true,
	}

	if policy.Defaults != nil {
		result.Decision = policy.Defaults.Decision
		result.Reason = policy.Defaults.Reason
	}

	return result
}

// ruleMatches checks if a rule matches the given context.
// All specified constraints must match (AND logic).
func ruleMatches(rule *PolicyRule, context *EvaluationContext) bool {
	// Check subject matcher
	if rule.Subject != nil && !matchesSubject(context.Subject, rule.Subject) {
		return false
	}

	// Check purpose
	if len(rule.Purpose) > 0 && !matchesPurpose(context.Purpose, rule.Purpose) {
		return false
	}

	// Check licensing mode
	if len(rule.LicensingMode) > 0 && !matchesLicensingMode(context.LicensingMode, rule.LicensingMode) {
		return false
	}

	return true
}

// matchesSubject checks if a subject matches the given matcher.
func matchesSubject(subject *Subject, matcher *SubjectMatcher) bool {
	if subject == nil {
		// If there's a subject matcher but no subject in context, no match
		// unless the matcher has no constraints
		return matcher.Type == "" && len(matcher.Labels) == 0 && matcher.ID == ""
	}

	// Check type
	if matcher.Type != "" && subject.Type != matcher.Type {
		return false
	}

	// Check labels - subject must have ALL required labels
	if len(matcher.Labels) > 0 {
		if !hasAllLabels(subject.Labels, matcher.Labels) {
			return false
		}
	}

	// Check ID pattern
	if matcher.ID != "" && !matchesIDPattern(subject.ID, matcher.ID) {
		return false
	}

	return true
}

// hasAllLabels checks if subjectLabels contains all required labels.
func hasAllLabels(subjectLabels []string, requiredLabels []string) bool {
	if len(requiredLabels) == 0 {
		return true
	}
	if len(subjectLabels) == 0 {
		return false
	}

	// Create a set of subject labels for O(1) lookup
	labelSet := make(map[string]bool, len(subjectLabels))
	for _, label := range subjectLabels {
		labelSet[label] = true
	}

	// Check all required labels are present
	for _, required := range requiredLabels {
		if !labelSet[required] {
			return false
		}
	}
	return true
}

// matchesIDPattern checks if an ID matches a pattern.
// Pattern can be exact match or prefix match with * suffix.
func matchesIDPattern(id string, pattern string) bool {
	if pattern == "" {
		return true
	}

	// Check for wildcard prefix match
	if strings.HasSuffix(pattern, "*") {
		prefix := strings.TrimSuffix(pattern, "*")
		return strings.HasPrefix(id, prefix)
	}

	// Exact match
	return id == pattern
}

// matchesPurpose checks if a purpose matches any of the allowed purposes.
func matchesPurpose(purpose ControlPurpose, allowed Purposes) bool {
	if len(allowed) == 0 {
		return true // No constraint means any purpose
	}
	if purpose == "" {
		return false // If allowed is specified but context has no purpose, no match
	}

	for _, p := range allowed {
		if p == purpose {
			return true
		}
	}
	return false
}

// matchesLicensingMode checks if a mode matches any of the allowed modes.
func matchesLicensingMode(mode ControlLicensingMode, allowed LicensingModes) bool {
	if len(allowed) == 0 {
		return true // No constraint means any mode
	}
	if mode == "" {
		return false // If allowed is specified but context has no mode, no match
	}

	for _, m := range allowed {
		if m == mode {
			return true
		}
	}
	return false
}

// IsAllowed returns true if the policy allows the context.
func IsAllowed(policy *PolicyDocument, context *EvaluationContext) bool {
	return Evaluate(policy, context).Decision == Allow
}

// IsDenied returns true if the policy denies the context.
func IsDenied(policy *PolicyDocument, context *EvaluationContext) bool {
	return Evaluate(policy, context).Decision == Deny
}

// RequiresReview returns true if the policy requires review for the context.
func RequiresReview(policy *PolicyDocument, context *EvaluationContext) bool {
	return Evaluate(policy, context).Decision == Review
}

// EvaluateBatch evaluates a policy against multiple contexts.
func EvaluateBatch(policy *PolicyDocument, contexts []*EvaluationContext) []*EvaluationResult {
	results := make([]*EvaluationResult, len(contexts))
	for i, ctx := range contexts {
		results[i] = Evaluate(policy, ctx)
	}
	return results
}
